GH-465: Clarify backward-compatibility rules on LIST type (#466)
apache/parquet-format: LogicalTypes.md

by `LIST` or `MAP` should be interpreted as a required list of required
elements where the element type is the type of the field.

> ```

Reviewer: I would suggest putting these examples, and the paragraph above, in a separate subsection, for example:
````markdown
Implementations should use either `LIST` and `MAP` annotations _or_ unannotated
repeated fields, but not both. When using the annotations, no unannotated
repeated types are allowed.

### Unannotated repeated fields

A repeated field that is neither contained by a `LIST`- or `MAP`-annotated
group nor annotated by `LIST` or `MAP` should be interpreted as a
required list of required elements where the element type is the type of
the field.

```
// List<Integer> (non-null list, non-null elements)
repeated int32 num;

// List<Tuple<Integer, String>> (non-null list, non-null elements)
repeated group my_list {
  required int32 num;
  optional binary str (STRING);
}
```

It is heavily recommended that writers use either the LIST or MAP annotations,
as outlined below, to make the intent clearer to readers.
````

Reviewer: I agree that it's slightly odd, but we would otherwise need a separate doc for interpretation of schemas as you say (and move any relevant contents out of `LogicalTypes.md`).

Author: It looks strange to make it a subsection if the unannotated repeated fields are not deprecated. If we deprecate it, `Backward-compatibility rules` section can be organized as below:
- 3-level structure with different field names
- 2-level structure
- Unannotated repeated fields

It would be better to deprecate it in another change. Therefore I tend to keep it as is.

Reviewer 2: > I'm not sure either why `parquet-rs` chosen to write "1-level" lists.

It doesn't by default. It takes user intervention to pass a schema with 1-level lists to the record writer. The table writer will use the proper 3-level structure AFAICT.

Reviewer 3: Thanks, @etseidl.

Related to protobuf (the parquet-java binding for protobuf) this is the other way around currently. There is a flag for converting repeated protobuf elements to "3-level" lists but it is false by default, so it converts directly to similar repeated elements in the parquet schema ("1-level" list).
Anyways it seems that neither parquet-rs nor parquet-java (protobuf) should have an issue creating proper "3-level" lists.

by `LIST` or `MAP` should be interpreted as a required list of required
elements where the element type is the type of the field.

```
// List<Integer> (non-null list, non-null elements)
repeated int32 num;

> // List<Tuple<Integer, String>> (non-null list, non-null elements)

Reviewer 4: nit: `List<Tuple<Integer, String>>` have some way to repr the `String` is nullable in type annotation?

repeated group my_list {
  required int32 num;
  optional binary str (STRING);

> }

Reviewer 5: I think this example is counter-productive. We don't want anyone using un-annotated lists and maps. While the paragraph above explains how to interpret un-annotated `repeated` fields, I don't want anyone to see an example here and think that it is something that should be copied. I think it is already clear enough and I would simply move on rather than drawing attention to this as a possibility.

Author: That make sense. Let me remove these examples first. I think a followup is to deprecate it by moving it to the backward compatibility section and adding strong words to discourage writers to emit it.

```

> For all fields in the schema, implementations should use either `LIST` and

Author: Explicitly state that this applies to `all fields in the schema`.

Reviewer 4: 👍I think this is neccessary, hope implemetation would follow this rule...

Reviewer 5: -0 on this change. I don't think this is more clear and I would prefer not to have the churn.

Author: I think the original sentence is confusing and resulted in a relevant discussion: https://github.com/apache/parquet-format/pull/466#pullrequestreview-2432448521. Therefore I prefer to keep the current change.

}
```

> Some existing data does not include the inner element layer, meaning that

Author: I have added more explanation according to the ML discussion. I'm not a native speaker, please help me check if these are suitable. Thanks! @rdblue @gszadovszky

> `LIST` annotates a 2-level structure. In contrast to 3-level structure, The

Reviewer 2: ```suggestion
`LIST` annotates a 2-level structure. In contrast to the 3-level structure, the
```

repetition of the outer level `LIST`-annotated 2-level structure can be

> `optional`, `required`, or `required`. For backward-compatibility, the type of

Reviewer 2: Is this one too many `required`? Or do you mean to add `repeated` as well?

elements in `LIST`-annotated structures should always be determined by the
following rules:

1. If the repeated field is not a group, then its type is the element type and
   elements are required. In this case, `LIST` annotates a 2-level structure.
2. If the repeated field is a group with multiple fields, then its type is the
   element type and elements are required. In this case, `LIST` annotates a
   2-level structure and the element type is a Struct type with multiple fields.
3. If the repeated field is a group (not `LIST`-annotated) with one `required`
   or `optional` field, and is named either `array` or uses the `LIST`-annotated
   group's name with `_tuple` appended, then the repeated type (a single-field
   Struct type) is the element type and elements are required. This is a special
   case of 3-level structure where the names are respected.
4. Otherwise, the repeated field's type is the element type with the repeated
   field's repetition. In this case, `LIST` annotates a 2-level structure. Note

>    that the repeated field cannot be a 3-level LIST whose repetition must be

Reviewer 2: ```suggestion
   that the repeated field cannot be a 3-level LIST, as such a list's repetition must be
```

}
```

##### 2-level structure

Some existing data does not include the inner element layer, meaning that `LIST`
annotates a 2-level structure. In contrast to 3-level structure, the repetition

> of 2-level structure can be `optional`, `required`, or `repeated`.

Reviewer 5: I find this confusing. Why does it need to change?

Author: I just happen to see that this section is all about 2-level list so I have added a title to make it explicit.

> The problem is when we need to decide whether a structure is a 2-level or 3-level.

The change aims to exhaust all possible forms of list structures and we can just apply these rules in order to deduce a list type:
- regular 3-level list
- 3-level list (w/ different names)
- 2-level list
- 1-level list (w/o list annotation)

Author: I decide to keep the subtitles as they are pretty clear. It is the exact order to apply when deducing a list type. Please let me know if you have different opinions and I'm happy to change. @rdblue @gszadovszky @emkornfield

```
<list-repetition> group <name> (LIST) {
  repeated <element-type> <element-name>;
}
```

For backward-compatibility, the type of elements in `LIST`-annotated 2-level
structures should always be determined by the following rules:

1. If the repeated field is not a group, then its type is the element type and
   elements are required.
2. If the repeated field is a group with multiple fields, then its type (Struct

>    type with multiple fields) is the element type and elements are required.

Reviewer 5: What is a "struct"? Do you mean the group with multiple fields? If so, then this is redundant isn't it?

Author: Yes, my intention is to differentiate the point 3 below which is a struct with a single field. I am remove the redundant statement if it is clear enough.

> 3. If the repeated field is a group with one `required` or `optional` field,

Reviewer 5: This rule should not be modified with new cases. I think we need to insert a new rule that states if the repeated field is a group with a `repeated` field, the repeated field is the element type because the type cannot be a 3-level list.

The rules should be as independent as possible and I think it makes it more confusing to mix them together like this and like the new #4.

   and is named either `array` or uses the `LIST`-annotated group's name with
   `_tuple` appended, then the repeated type (Struct type with single field) is
   the element type and elements are required.
4. If the repeated field is a `LIST`-annotated group with one `repeated` field,
   then the element type is a list type with 2-level structure and elements are
   required.
5. Otherwise, the repeated field's type is the element type with the repeated
   field's repetition. Please note that the repeated field here (a group with
   one field) cannot be `LIST`-annotated or `MAP`-annotated 3-level structure,
   as such a group's repetition must be `required` or `optional`.

Examples that can be interpreted using these rules:

```
// Rule 1: List<Integer> (nullable list, non-null elements)
optional group my_list (LIST) {
  repeated int32 element;
}

// Rule 2: List<Struct<String, Integer>> (nullable list, non-null elements)
optional group my_list (LIST) {
  repeated group element {
    required binary str (STRING);
    required int32 num;
  };
}

> // Rule 3: List<Struct<String>> (nullable list, non-null elements)

Reviewer 5: I think this makes the example less clear.

optional group my_list (LIST) {
  repeated group array {
    required binary str (STRING);
  };
}

// Rule 3: List<Struct<String>> (nullable list, non-null elements)
optional group my_list (LIST) {
  repeated group my_list_tuple {
    required binary str (STRING);
  };
}

// List<List<Integer>>
// Rule 4: nullable outer list with non-null elements
// Rule 1: non-null inner list with non-null elements
optional group my_list (LIST) {
  repeated group array (LIST) {
    repeated int32 array;
  };
}

// Rule 5: List<Struct<List<Integer>>> (nullable outer list with non-null elements)
optional group my_list (LIST) {
  repeated group foo {
    repeated int32 bar;
  };
}
```

##### 1-level structure without `LIST` annotation

Some existing data does not even have the `LIST` annotation and simply uses

> `repeated` repetition to annotate the element type. In this case, the element
> type MUST be a primitive type and both the list and elements are required.

Reviewer 2: ```suggestion
`repeated` repetition to annotate the element type. In this case both the list
and elements are required.
```

}
```

##### 2-level structure

Some existing data does not include the inner element layer, meaning that `LIST`
annotates a 2-level structure. In contrast to 3-level structure, the repetition
of 2-level structure can be `optional`, `required`, or `repeated`.

```
<list-repetition> group <name> (LIST) {
  repeated <element-type> <element-name>;
}
```

For backward-compatibility, the type of elements in `LIST`-annotated 2-level
structures should always be determined by the following rules:

1. If the repeated field is not a group, then its type is the element type and

>    elements are required.

Reviewer 4: A unrelated: I just find this "elements are required" is a bit tricky for me...

2. If the repeated field is a group with multiple fields, then its type (Struct
   type with multiple fields) is the element type and elements are required.

> 3. If the repeated field is a group with one `required` or `optional` field,

Reviewer 4: with only one?

   and is named either `array` or uses the `LIST`-annotated group's name with
   `_tuple` appended, then the repeated type (Struct type with single field) is
   the element type and elements are required.

> 4. If the repeated field is a `LIST`-annotated group with one `repeated` field,

Author: No, rule 3 is for a group with one `required` or `optional` field while rule 4 is for a `LIST`-annotated group with one `repeated` field. They do not overlap and the rule 3 implies that the group cannot have any annotation.

> 4. If the repeated field is a `LIST`-annotated group with one `repeated` field,

Reviewer 2: Got it, thanks.

   then the element type is a list type with 2-level structure and elements are
   required.
5. Otherwise, the repeated field's type is the element type with the repeated
   field's repetition. Please note that the repeated field here (a group with
   one field) cannot be `LIST`-annotated or `MAP`-annotated 3-level structure,
   as such a group's repetition must be `required` or `optional`.

Examples that can be interpreted using these rules:

```
// Rule 1: List<Integer> (nullable list, non-null elements)
optional group my_list (LIST) {
  repeated int32 element;
}

// Rule 2: List<Struct<String, Integer>> (nullable list, non-null elements)
optional group my_list (LIST) {
  repeated group element {
    required binary str (STRING);
    required int32 num;
  };
}

// Rule 3: List<Struct<String>> (nullable list, non-null elements)
optional group my_list (LIST) {
  repeated group array {
    required binary str (STRING);
  };
}

// Rule 3: List<Struct<String>> (nullable list, non-null elements)
optional group my_list (LIST) {
  repeated group my_list_tuple {
    required binary str (STRING);
  };
}

> 

Reviewer 4: do we have sample for Rule 5?

Author: As I have explained in https://lists.apache.org/thread/s6b25j3x26009v054yqjov0f1z49ctqj, the only case left for rule 5 is as below:
```
optional group foo (LIST) {
  repeated group bar {
    repeated TYPE baz;
  };
}
```

It should be resolved to `List<Struct<List<TYPE>>>` if the 1-level structure is allowed, or an invalid case at all.

// List<List<Integer>>
// Rule 4: nullable outer list with non-null elements
// Rule 1: non-null inner list with non-null elements
optional group my_list (LIST) {
  repeated group array (LIST) {
    repeated int32 array;
  }
}
```

##### 1-level structure without `LIST` annotation

Some existing data does not even have the `LIST` annotation and simply uses
`repeated` repetition to annotate the element type. In this case, the element

> type MUST be a primitive type and both the list and elements are required.

Reviewer 2: I don't know if this belongs in the backwards compatibility section. AFAICT using `repeated` without the `LIST` annotation is still supported by the spec (and some writers can still produce it, including arrow-rs; cc @zeevm).

Also, I don't think the current [wording](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#nested-types) requires a primitive type. The following should be allowed (`List<Struct<Integer, String>> non-null list, non-null elements`):
```
repeated group list_struct {
  required int32 id;
  optional binary val (STRING);
}
```
In fact, in parquet-testing there's a file `repeated_no_annotation.parquet` with an unannotated `repeated` group:
```
message user {
  REQUIRED INT32 id;
  OPTIONAL group phoneNumbers {
    REPEATED group phone {
      REQUIRED INT64 number;
      OPTIONAL BYTE_ARRAY kind (UTF8);
    }
  }
}
```

Author: > AFAICT using repeated without the LIST annotation is still supported by the spec

I don't think it is supported by the spec because it is unclear yet. The official list type in the spec is the LIST-annotated group with 3-level structure, which support arbitrary nesting and full capability to specify nullability of each level. The LIST-annotated group with 3-level structure should always be used by writers and others fall into the category of backward compatibility to deal with existing files. A writer can accidentally produce such files does not mean it should be that way.

> Also, I don't think the current [wording](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#nested-types) requires a primitive type

That's true. Let me change this.

1. If the repeated field is not a group, then its type is the element type and
   elements are required.
2. If the repeated field is a group (without `LIST` or `MAP` annotation) with
   multiple fields, then its type (i.e. a multi-field Tuple) is the element
   type and elements are required.
3. If the repeated field is a group (without `LIST` or `MAP` annotation) with
   one field (with `required` or `optional` repetition) and is named either
   `array` or uses the `LIST`-annotated group's name with `_tuple` appended,
   then the repeated type is the element type and elements are required. Note
   that the element type resolves to a single-field Tuple type.
4. Otherwise, the repeated field's type is the element type with the repeated
   field's repetition and elements are required. Note that the element type
   resolves to List or Map type if the repeated field is `LIST`-annotated or

>    `MAP`-annotated; otherwise the element type resolves to a Tuple type.

Reviewer 4: I don't know the worlding of "Tuple type" is better or "Struct type" is better...

Author: The example below uses `Tuple` and `OneTuple` and we'd better be consistent.

1. If the repeated field is not a group, then its type is the element type and
   elements are required.
2. If the repeated field is a group with multiple fields, then its type is the
   element type and elements are required. In this case, the element type is
   a Struct type with multiple fields.

> 3. If the repeated field is a group (without annotation) with one `required` or

Reviewer 6: I might leave the entire block as it was previously and clarify that:
1. Without annotatation the type is a single struct/tuple with the element in it.
2. With a logical type annotation (I think only LIST and MAP apply here).

I also think we need to update the documentation above:
```
The outer-most level must be a group annotated with `LIST` that contains a
  single field named `list`. The repetition of this level must be either
  `optional` or `required` and determines whether the list is nullable.
```
To point to the fact that for backwards compatibility it can also be repeated.

Author: > With a logical type annotation (I think only LIST and MAP apply here).

After double check, I don't think MAP annotation will appear here because it always has three levels.

> To point to the fact that for backwards compatibility it can also be repeated.

The original statement is all about the formal three-level encoding. Adding `repeated` to it may further confuse readers. That's why I tried to clarify this in the backward compatibility section.

> 3. If the repeated field is a group (without annotation) with one `required` or

Author: I doesn't hurt to make it explicit.

> I'm also wondering if we should add an example above for the non-LIST annotated repeating fields, or should that be a new PR?

I think we should, but I am not clear where to add it since these lines are all talking about LIST-annotated structures.

> 3. If the repeated field is a group (without annotation) with one `required` or

Reviewer 2: > I think we should, but I am not clear where to add it since these lines are all talking about LIST-annotated structures.

Perhaps around line 575 where the unannotated form is introduced. 🤷  I admit this is a bit of an awkward location, since this file is about logical types. Out of scope here, but perhaps a new markdown file that documents the various ways repetition is supported would be appropriate. I know it would help me.

> 3. If the repeated field is a group (without annotation) with one `required` or

Author: I think we'd better open another PR to discuss list types without annotation. It might be more complicated than the scope of this PR. I'm not sure if it should be allowed by the spec even though we are able to create the schema.

#### Backward-compatibility rules

> Modern writers should always produce the 3-level LIST structure shown above.

Reviewer 6: ```suggestion
New writer implementations should always produce the 3-level LIST structure shown above.
```

> #### Backward-compatibility rules

Reviewer: It is nice that we are including rules for how to interpret existing data, but can we also have guidelines for writers such that we converge on a single rule for future data? It would be nice to reduce the existing variability in the long term.

Reviewer 3: The current spec (and it didn't change with this PR) specifies the "3-level" structure and the "unannotated" (aka. "1-level") list types as ones to be supported. So, according to the spec, a writer either use the first or the second one but cannot use both in the same schema.

Since only the "3-level" structure is capable of properly describing nullability at any level, this should be the preferred one.

We may start a discussion to deprecate the "1-level" list but I think this would be a different topic. This PR is about clarifying the backward compatibility rules only.

Reviewer: Okay, so I've suggested an addition below to make this extra clear.

> 

Reviewer: ```suggestion

Modern writers should always produce the 3-level LIST structure shown above.
However, historically data files have been produced that use different structures to
represent list-like data, and readers may include compatibility measures to interpret
them as intended.

```

    repeated int32 array;
  };
}

// Rule 4: List<OneTuple<String>> (nullable list, non-null elements)
optional group my_list (LIST) {
  repeated group array {
    required binary str (STRING);
  };
}

// Rule 4: List<OneTuple<String>> (nullable list, non-null elements)
optional group my_list (LIST) {
  repeated group my_list_tuple {
    required binary str (STRING);
  };
}

// Rule 5: List<OneTuple<List<Integer>>> (nullable outer list, non-null elements)
optional group my_list (LIST) {
  repeated group foo {
    repeated int32 bar;
  };
}
```

##### 1-level structure without `LIST` annotation

Some existing data does not even have the `LIST` annotation and simply uses
`repeated` repetition to annotate the element type. For backward-compatibility,
both the list and elements are `required`.

```
// List<Integer> (non-null list, non-null elements)
repeated int32 num;

// Tuple<List<Integer>, List<String>> (non-null list, non-null elements)
optional group my_list {
  repeated int32 num;
  repeated binary str (STRING);
}

// List<Tuple<Integer, String>> (non-null list, non-null elements)
repeated group my_list {
  required int32 num;
  optional binary str (STRING);
}

> ```

Reviewer 3: As @etseidl mentioned before, there is a [section](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md?plain=1#L604-L614) just before lists that clearly states that the repeated fields without annotation is supported. So, it is misleading that we explain the "1-level structure" at the backward compatibility section. Maybe move the examples to the existing section and keep only a reference that states the "1-level structure" is actually a 1st class citizen in the spec (and not deprecated).

Or actually deprecate it now? Iceberg seems to only support the 3-level list. Meanwhile, there are engines still write the 1-level one, like parquet-rs. Even parquet-java writes these in case of the protobuf binding is used.

Author: Good catch! I have moved the example closer to the description. IMO, we should not deprecate it because it is not confusing and actually much cleaner to present a non-null list.

Reviewer 3: It might be cleaner but not sure if we can safely mix them. Even if we can, it contradicts to the spec. See [here](https://github.com/apache/parquet-format/pull/466#pullrequestreview-2432448521).

}
```

Some existing data does not include the inner element layer, resulting in a
`LIST` that annotates a 2-level structure. Unlike the 3-level structure, the
repetition of a 2-level structure can be `optional`, `required`, or `repeated`.
When it is `repeated`, the `LIST`-annotated 2-level structure can only serve as
an element within another `LIST`-annotated 2-level structure.

For backward-compatibility, the type of elements in `LIST`-annotated structures
should always be determined by the following rules:

1. If the repeated field is not a group, then its type is the element type and
   elements are required.
2. If the repeated field is a group with multiple fields, then its type is the
   element type and elements are required.
3. If the repeated field is a group with one field with `repeated` repetition,
   then its type is the element type and elements are required.
4. If the repeated field is a group with one field and is named either `array`
   or uses the `LIST`-annotated group's name with `_tuple` appended then the
   repeated type is the element type and elements are required.

> 5. Otherwise, the repeated field's type is the element type with the repeated

Author: I have reverted most of the previous changes and now it should be clear. @etseidl @mapleFU 

To resolve a LIST-annotated group, we should apply rules in order:
- check if it is a 2-level structure (rule 1 to 3)
- check if it is a special 2-level structure (rule 4)
- it is a 3-level structure (rule 5)

Author: I don't want to add an example for rule 5 because it is already at line 685

Author: @pitrou Regarding to your [question](https://github.com/apache/parquet-format/pull/466#discussion_r1868965095) on rule 5, I think it is redundant to add another one.

Reviewer: Redundant with what? We've been explicitly listing examples for all other rules.

Author: Ok, I added one with a slightly different schema. The inner field is `optional` in the new example. Please take a look.

Reviewer: Ok, thanks.

}
```

For backward-compatibility, the type of elements in `LIST`-annotated 2-level
structures should always be determined by the following rules:

1. If the repeated field is not a group, then its type is the element type and
   elements are required.
2. If the repeated field is a group with multiple fields, then its type is the
   element type and elements are required.
3. If the repeated field is a group with a `repeated` field, then the repeated
   field is the element type because the type cannot be a 3-level list.
4. If the repeated field is a group with one field and is named either `array`
   or uses the `LIST`-annotated group's name with `_tuple` appended then the
   repeated type is the element type and elements are required.
5. Otherwise, the repeated field's type is the element type with the repeated
   field's repetition.

Examples that can be interpreted using these rules:

```
// Rule 1: List<Integer> (nullable list, non-null elements)
optional group my_list (LIST) {
  repeated int32 element;
}

// Rule 2: List<Tuple<String, Integer>> (nullable list, non-null elements)
optional group my_list (LIST) {
  repeated group element {
    required binary str (STRING);

>     required int32 num;

Reviewer: But this child field could be "optional", right?

Reviewer 2: `str` and `num` can be optional, but the `element` itself cannot be null (but `my_list` can be empty).

Reviewer 4: Below there is a similiar example with `optional str` after `num`

  };
}

// Rule 3: List<List<Integer>> (nullable outer list, non-null elements)
optional group my_list (LIST) {
  repeated group array (LIST) {
    repeated int32 array;
  };
}

// Rule 4: List<OneTuple<String>> (nullable list, non-null elements)
optional group my_list (LIST) {
  repeated group array {
    required binary str (STRING);
  };
}

> // Rule 4: List<OneTuple<String>> (nullable list, non-null elements)

Reviewer 3: `Rule 4` is only about naming and these are the two options. I think it is fine to have both to be clear.

> // Rule 4: List<OneTuple<String>> (nullable list, non-null elements)

Reviewer: What would happen if the naming does not satisfy Rule 4? We should give an example of this fallback case.

(I'm also afraid this rule is pointlessly complicated)

> // Rule 4: List<OneTuple<String>> (nullable list, non-null elements)

Author: If naming does not satisfy Rule 4, then Rule 5 applies (though it produces the same schema).

```
// Rule 4: List<OneTuple<String>> (nullable list, non-null elements)
optional group my_list (LIST) {
  repeated group array {
    required binary str (STRING);
  };
}

// Rule 4: List<OneTuple<String>> (nullable list, non-null elements)
optional group my_list (LIST) {
  repeated group my_list_tuple {
    required binary str (STRING);
  };
}

// Rule 5: List<OneTuple<String>> (nullable list, non-null elements)
optional group my_list (LIST) {
  repeated group not_array_or_my_list_tuple {
    required binary str (STRING);
  };
}
```

I think @rdblue have suggested not to change this rule for historical reason. Now it is pretty clear that these legacy namings do not make a difference, should we merge the last two rules to ignore their names?

> // Rule 4: List<OneTuple<String>> (nullable list, non-null elements)

Reviewer 2: But now the Rule 5 example is simply a 3-level list with different names. It would be `List<String>`.

optional group my_list (LIST) {
  repeated group my_list_tuple {
    required binary str (STRING);
  };
}

// Rule 5: List<OneTuple<List<Integer>>> (nullable outer list, non-null elements)

> optional group my_list (LIST) {

Author: I believe this is the case where annotated 2-level list is mixed with unannotated repeated field (1-level list). Is this allowed? If not, I can remove the example. Then, is there any concrete example for the last rule (a.k.a. rule 5 above)? @gszadovszky @rdblue

Reviewer 2: If my comment above is correct, perhaps keep the Rule 5 example, but instead explain why it is not allowed.

Author: @etseidl Your analysis aligns with mine on the dev ML. I don't think there is any real case in rule 5.

}
```

##### 2-level structure

Some existing data does not include the inner element layer, resulting in a
`LIST` that annotates a 2-level structure. Unlike the 3-level structure, the
repetition of a 2-level structure can be `optional`, `required`, or `repeated`.
When it is `repeated`, the `LIST`-annotated 2-level structure can only serve as
an element within another `LIST`-annotated 2-level structure.

```
<list-repetition> group <name> (LIST) {
  repeated <element-type> <element-name>;

> }

Reviewer 5: Again, I think that calling attention to the degenerate cases and documenting them is only going to cause more confusion. The purpose of this originally was to simply document how to interpret data that doesn't match expectations. Now this introduces how a 2-level list looks, which I think increases the possibility that people will misread this and write them.

Author: Removed

```

For backward-compatibility, the type of elements in `LIST`-annotated structures
should always be determined by the following rules if they cannot be determined
as 3-level structures:

1. If the repeated field is not a group, then its type is the element type and
   elements are required.
2. If the repeated field is a group with multiple fields, then its type is the
   element type and elements are required.
3. If the repeated field is a group with one field with `repeated` repetition,
   then its type is the element type and elements are required.

> 4. If the repeated field is a group with one field and is named either `array`

Reviewer 4: So, the rule-4 doesn't means like (3), but inner-most element type is "required" ( not repeated ?)

Author: The inner-most field is repeated in rule 3 but cannot be repeated in rule 4. In addition, rule 4 refers to a special case produced by legacy thrift and avro writer in the parquet-mr, therefore it must use the exact field name to match.

   or uses the `LIST`-annotated group's name with `_tuple` appended then the
   repeated type is the element type and elements are required.
5. Otherwise, the repeated field's type is the element type with the repeated
   field's repetition.

Examples that can be interpreted using these rules:

```
// Rule 1: List<Integer> (nullable list, non-null elements)
optional group my_list (LIST) {
  repeated int32 element;
}

// Rule 2: List<Tuple<String, Integer>> (nullable list, non-null elements)
optional group my_list (LIST) {
  repeated group element {
    required binary str (STRING);
    required int32 num;
  };
}

// Rule 3: List<List<Integer>> (nullable outer list, non-null elements)
optional group my_list (LIST) {
  repeated group array (LIST) {

>     repeated int32 array;

Reviewer 4: Still cannot figure out rule3 and rule4 diffrences...

Author: See my comment above.

}
```

##### 2-level structure

Some existing data does not include the inner element layer, resulting in a
`LIST` that annotates a 2-level structure. Unlike the 3-level structure, the

> repetition of a 2-level structure can be `optional`, `required`, or `repeated`.

Reviewer 4: "the repetition of a 2-level structure can be" so here means the outer "list-repetition"? Should we denote it as out-most level?

Author: I think this is clear because we are talking about the whole 2-level structure here.

When it is `repeated`, the `LIST`-annotated 2-level structure can only serve as
an element within another `LIST`-annotated 2-level structure.

```
<list-repetition> group <name> (LIST) {
  repeated <element-type> <element-name>;
}
```

> For backward-compatibility, the type of elements in `LIST`-annotated structures

Author: Explicitly add `if they cannot be determined as 3-level structures` to partially address the comment on how to resolve a list-annotated group in practice.

Reviewer 4: So here following the rule:
1. Trying to parse as 3-level structure
2. Trying to parse 2-level legacy
3. Trying to parse 1-level legacy

?

Author: I agree with @etseidl. This section only discusses LIST-annotated structures.

should always be determined by the following rules if they cannot be determined
as 3-level structures:

1. If the repeated field is not a group, then its type is the element type and
   elements are required.
2. If the repeated field is a group with multiple fields, then its type is the
   element type and elements are required.
3. If the repeated field is a group with one field and the repetition of that

>    field is `repeated`, then its type is the element type and elements are

Reviewer 4: I fount the "repeated field" is `repeated` is ambigious here...

Author: Changed the sentence a little bit.

   repeated type is the element type and elements are required.
5. Otherwise, the repeated field's type is the element type with the repeated
   field's repetition.

Examples that can be interpreted using these rules:

```
WARNING: writers should not produce list types like these examples! They are
just for the purpose of reading existing data for backward-compatibility.

// Rule 1: List<Integer> (nullable list, non-null elements)
optional group my_list (LIST) {
  repeated int32 element;
}

// Rule 2: List<Tuple<String, Integer>> (nullable list, non-null elements)
optional group my_list (LIST) {
  repeated group element {
    required binary str (STRING);
    required int32 num;
  };
}

// Rule 3: List<List<Integer>> (nullable outer list, non-null elements)
optional group my_list (LIST) {
  repeated group array (LIST) {
    repeated int32 array;
  };
}

// Rule 4: List<OneTuple<String>> (nullable list, non-null elements)
optional group my_list (LIST) {
  repeated group array {
    required binary str (STRING);
  };
}

// Rule 4: List<OneTuple<String>> (nullable list, non-null elements)
optional group my_list (LIST) {
  repeated group my_list_tuple {

>     required binary str (STRING);

Reviewer: For the record: I would expect an example of Rule 5 below?
